### PR TITLE
Add application guidelines for 1.0

### PR DESCRIPTION
**Pull request by texasmichelle**
Moving from Gdoc draft

/cc @jlewi 
/cc @ddutta 
/cc @kkasravi 
/cc @abhi-g

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/community/285)

### CODE REVIEW

**Comment by jlewi** [comment]
Thanks @texasmichelle !

@richardsliu @jbottum @johnugeorge @vkoukis  could you take a look please?

File: guidelines/application_requirements.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+1. Custom resources and controllers
+1. Web applications
+
+This document is not intended to cover SDKs (e.g. fairing) or client side tools (like kfctl).
+
+## Goals
+
+The purpose of this doc is to:
+
+1. Establish criterion for graduating a Kubeflow application to a supported Kubeflow application; i.e. 1.0
```
Comment (sarahmaddox) [inline guidelines/application_requirements.md]: s/criterion/criteria

(Because "criteria" is the plural of "criterion".)

Comment (sarahmaddox) [inline guidelines/application_requirements.md]: From discussions earlier today: When we talk about moving a Kubeflow application to v1.0, there are actually two things involved:
* Moving the application itself to v1.0 of the application.
* Declaring the application to be a supported part of Kubeflow v1.x (or 2.x, etc).

I think we should make this clear in this doc, and also clarify whether these are the criteria for the app itself to move to v1.0 or for the app to be considered "Kubeflow 1.0+ compliant".

File: guidelines/application_requirements.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+1. Set clear expectations for users regarding Kubeflow applications that are 1.0
```
Comment (sarahmaddox) [inline guidelines/application_requirements.md]: Same here - are we setting expectations for applications that are 1.0, or those that are compliant with Kubeflow 1.0+ standards?

File: guidelines/application_requirements.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+| Description | Category | Explanation |
+|-------------|----------|-------------|
+| List of users running the application | Recommended  | <ul><li>Suggest listing adopters willing to be identified publicly in ADOPTERS.md </ul> |
+
+## Reference
+
+* TFJob and PyTorch 1.0 [exit criterion](http://bit.ly/operators-exit-criterion)
+* [CNCF Graduation Criteria](https://github.com/cncf/toc/blob/master/process/graduation_criteria.adoc)
+* [Core Infrastructure Best Practices](https://github.com/coreinfrastructure/best-practices-badge)
+* [v1.0 Docs and Processes](https://docs.google.com/document/d/1v06QmjIms3z-uoW-waS7S9IzMA6n4CVrXOwQQ9N0do4)
```
Comment (sarahmaddox) [inline guidelines/application_requirements.md]: s/v1.0/Kubeflow v1.0

File: guidelines/application_requirements.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+## Requirements
+
+### Configuration and deployment
+
+| Description | Category | Explanation |
+|-------------|----------|-------------|
+| Kustomize package  | Required  | <ul><li>Kubeflow has standardized on Kustomize for configuring and deploying packages </ul> |
+| Application CR  | Required  | <ul><li>Kubeflow has standardized on using the Kubernetes Application CR to provide consistent metadata and monitoring about applications <li> Application CR should be an owner of all resources so that deleting the application CR uninstalls the application </ul> |
+| ```app.kubernetes.io``` labels on every resource | Required  | <ul><li>Every resource that is part of the application should include the [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels) recommended by Kubernetes </ul> |
```
Comment (johnugeorge) [inline guidelines/application_requirements.md]: Give more information on labels to be added to ensure consistency? 

eg: https://github.com/kubeflow/manifests/blob/master/tf-training/tf-job-operator/overlays/application/application.yaml

**Comment by richardsliu** [comment]
Can we include some general guidelines for custom resources?

Ex:
- Version stability (v1, no deprecative API changes)
- Backward compatibility (serves multiple versions)
- Supports subresource status
- Supports CRD schema validation
- For training operators, follow the conventions defined in kubeflow/common (https://github.com/kubeflow/common/blob/master/job_controller/api/v1/types.go)
- ??

File: guidelines/application_requirements.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+| Images listed in kustomization.yaml | Required  | <ul><li>All docker images used by the application should be listed in one or more kustomization.yaml files <li>This facilitates mirroring the images to different docker registries </ul> |
+| Upgradeability | Required  | <ul><li>Applications must support upgrading between consecutive major and minor releases </ul> |
+| Separate cluster scoped and namespace scoped resources | Recommended  | <ul><li>To the extent possible cluster scoped resources should be installable separately (e.g. via a separate kustomize package) <li> This allows cluster admins to install only the cluster scoped resources </ul> |
```
Comment (johnugeorge) [inline guidelines/application_requirements.md]: In ksonnet, we had supported a flag that determined namespace scoped vs cluster scoped installation of operator. Currently, we don't do that in Kustomize which uses cluster role binding by default.   We have to document the right practices for writing manifests for each application.

Comment (johnugeorge) [inline guidelines/application_requirements.md]: It would be also better to document cluster scoped vs namespace scoped behavior that has to be supported in each application.  eg: KFServing operator, Katib operator don't support namespace scope while Tfjob/PyTorch Job supports namespace scope in its implementation.

**Comment by johnugeorge** [comment]
As suggested by @richardsliu , we need a consistent behavior across CRs.    Keeping TFJob/PyTorchJob as a reference is a good idea as it has graduated.

One such example is:  There are different requirements of minimum kubernetes versions across CRs in Kubeflow now. Status sub resource requires 1.11+.  Inorder to support multiple versions in Knative(KFserving dependency), it needs k8s 1.14+ (https://github.com/knative/serving/issues/4533)

**Comment by kkasravi** [comment]
> Can we include some general guidelines for custom resources?
> 
> Ex:
> 
> * Version stability (v1, no deprecative API changes)
> * Backward compatibility (serves multiple versions)
> * Supports subresource status
> * Supports CRD schema validation
> * For training operators, follow the conventions defined in kubeflow/common (https://github.com/kubeflow/common/blob/master/job_controller/api/v1/types.go)
> * ??


Richard - what sort of sub resource status were you thinking of?

**Comment by johnugeorge** [comment]
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource

File: guidelines/application_requirements.md (hunk 2 above)
Comment (ashahba) [inline guidelines/application_requirements.md]: maybe reword it to:
```
Define clear requirements for Kubeflow 1.0 application development along with end user docs and examples
```

File: guidelines/application_requirements.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+Authors: chasm@google.com, jlewi@google.com, kam.d.kasravi@intel.com
+
+## Objective
+
+Define requirements and recommendations aimed at graduating applications (e.g. controller and web apps) in Kubeflow to 1.0.
+
+## Scope
+
+This document is intended to cover applications that are deployed in the Kubernetes cluster and in particular:
+
```
Comment (kbhawkey) [inline guidelines/application_requirements.md]: nit: ... in which Kubernetes cluster? The cluster that also contains `kubeflow`?

File: guidelines/application_requirements.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+1. Custom resources and controllers
+1. Web applications
+
+This document is not intended to cover SDKs (e.g. fairing) or client side tools (like kfctl).
+
+## Goals
+
```
Comment (kbhawkey) [inline guidelines/application_requirements.md]: nit: The first item in `Goals` sounds very similar to the `Objective`. Do you need both or can you further define the goal?

File: guidelines/application_requirements.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+| Upgradeability | Required  | <ul><li>Applications must support upgrading between consecutive major and minor releases </ul> |
+| Separate cluster scoped and namespace scoped resources | Recommended  | <ul><li>To the extent possible cluster scoped resources should be installable separately (e.g. via a separate kustomize package) <li> This allows cluster admins to install only the cluster scoped resources <li> Clear documentation of cluster and namespace scoped behavior </ul> |
+| Kustomize package should be deployable on its own | Recommended  | <ul><li>To the extent possible users should be able to run kustomize build | kubectl apply inside the kustomize package to deploy the application stand alone </ul> |
+
+### Custom Resources
+
+| Description | Category | Explanation |
+|-------------|----------|-------------|
+| Version stability | Required  | <ul><li>No deprecative API changes </ul> |
+| Backward compatibility | Required  | <ul><li>Serves multiple versions </ul> |
```
[the remark this answers is not in the document]
Comment (johnugeorge) [reply guidelines/application_requirements.md]: Agree. That is more consistent with k8s terminology and gives more clarity.

**Comment by johnugeorge** [comment]
@texasmichelle We have seen issues where users have tried to install Kubeflow on older versions of K8s and it failed to deploy. 

Related:  https://github.com/kubeflow/pytorch-operator/issues/206

Hence it is better to list these requirements as all components are deployed together.

File: guidelines/application_requirements.md (hunk 5 above)
Comment (jlewi) [reply guidelines/application_requirements.md]: Apologies @richarddli I meant @richardsliu

File: guidelines/application_requirements.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (jlewi) [reply guidelines/application_requirements.md]: @texasmichelle I added a commit with changes to incorporate this feedback.

[the remark this answers is not in the document]
Comment (texasmichelle) [reply guidelines/application_requirements.md]: LGTM, thanks for the updates!

**Comment by jlewi** [comment]
@texasmichelle Thank you. I did another pass through the document; LGTM.

I think there's been ample time for folks to comment on this.

/lgtm
/approve

/hold for LGTM by @kkasravi and @ddutta 


@kkasravi and @ddutta can you please LGTM?
